gcip/addons/container/sequences/build.py

```diff
@@ -0,0 +1,97 @@
+
+
+def build_scan_push_image(
+    *,
+    registry: Union[Registry, str] = Registry.DOCKER,
+    image_name: Optional[str] = None,
+    image_tag: Optional[str] = None,
+    kaniko_kwargs: Dict[str, Any] = {},
+    dive_kwargs: Dict[str, Any] = {},
+    trivy_kwargs: Dict[str, Any] = {},
+    crane_kwargs: Dict[str, Any] = {},
+    docker_client_config: Optional[DockerClientConfig] = None,
+) -> Sequence:
+    """
+    Creates a `gcip.Sequence` to build, scan and push a container image.
+
+    The build step is executed by `gcip.addons.container.jobs.kaniko.execute`, it will build the container image an outputs it to a tarball.
+    There are two scan's, optimization scan with `gcip.addons.container.jobs.dive.scan_local_image` to scan storage wasting in container image
+    and a vulnerability scan with `gcip.addons.container.jobs.trivy.scan`. Both outputs are uploaded as an artifact to the GitLab instance.
+    Built container image is uploaded with `gcip.addons.container.jobs.crane.push`.
+
+    Args:
+        registry (Union[Registry, str], optional): Container registry to push the image to. If the container registry needs authentication,
+            you have to provide a `gcip.addons.container.config.DockerClientConfig` object with credentials. Defaults to Registry.DOCKER.
+        image_name (Optional[str]): Image name with namespace in the registry. e.g. username/image_name.
+            Defaults to `gcip.core.variables.PredefinedVariables.CI_PROJECT_NAME`.
+        image_tag (Optional[str]): Image tag. The default is either `PredefinedVariables.CI_COMMIT_TAG` or
+            `PredefinedVariables.CI_COMMIT_REF_SLUG` depending of building from a git tag or from a branch.
+        kaniko_kwargs (Dict[str, Any]): Extra keyword arguaments passed to `gcip.addons.container.jobs.kaniko.execute`. Defaults to {}.
+        dive_kwargs (Dict[str, Any], optional): Extra keyword arguaments passed to `gcip.addons.container.jobs.dive.scan`. Defaults to {}.
+        trivy_kwargs (Dict[str, Any], optional): Extra keyword arguaments passed to `gcip.addons.container.jobs.trivy.scan_local_image`. Defaults to {}.
+        crane_kwargs (Dict[str, Any], optional): Extra keyword arguaments passed to `gcip.addons.container.jobs.push`. Defaults to {}.
+        docker_client_config (Optional[DockerClientConfig], optional): Creates the Docker configuration file base on objects settings,
+            to authenticate against given registries. Defaults to a `DockerClientConfig` with login to the official Docker Hub
+            and expecting credentials given as environment variables `REGISTRY_USER` and `REGISTRY_LOGIN`.
+
+    Returns:
+        Sequence: `gcip.Sequence` to build, scan and push a container image.
+    """
+    cache = Cache(paths=["image"])
+    kaniko_execute = kaniko.execute(
+        image_name=image_name,
+        image_tag=image_tag,
+        registries=[registry],
+        tar_path=cache.paths[0],
```

Comment:
For now setting the `tar_path` ( and and euqivalent cache paths in later jobs) this way is ok and following suggestions should not block this PR to be merged.

However it would be much nicer, if a gcip job generator like `kaniko.execute()` also uses gcip features. When I am right, the only reason to set the `tar_path` ( or `image_path`) in one of those jobs, is using them for caching... maybe also for artifacts. So instead of just passing an 'arbitrary' path to this `tar_path` parameter, it could be more meaningfull to the user to pass a `Cache` (or `Artifacts`) object to this job. The job could then use for example the first path (should be documented) to store its output. Maybe the parameter could then also be named `cache` (or `tar_cache`, ...) instead of `tar_path`. 

The current way is not easy to reconstruct without good documentation, as you need to know to pass the `path[0]` component of a cache object to all those methods. It would be easier to just "throw in" an cache object. And going a step further - it would be much much cooler, if the first job `kaniko.execute` just emits a cache object request. So if you - for example - set the parameter `return_cache=True`, then the method could return that cache object, you could use for all subsequent container jobs. You then didn't bother about the Caches `paths=...` parameter.

Should I create an feature request Issue for it?

Reply:
https://github.com/dbsystel/gitlab-ci-python-library/issues/59